src/libaktualizr/primary/reportqueue.cc

```diff
@@ -32,17 +33,15 @@ void ReportQueue::run() {
 
 void ReportQueue::enqueue(std::unique_ptr<ReportEvent> event) {
   {
-    std::lock_guard<std::mutex> lock(m_);
-    report_queue_.push(std::move(event));
+    //std::lock_guard<std::mutex> lock(m_);
+    storage->saveReportEvent(event->toJson());
   }
   cv_.notify_all();
 }
 
 void ReportQueue::flushQueue() {
-  while (!report_queue_.empty()) {
-    report_array.append(report_queue_.front()->toJson());
-    report_queue_.pop();
-  }
+  int64_t max_id = 0;
+  storage->loadReportEvents(&report_array, &max_id);
```

Comment:
If you use load everything from storage here, you do not need to keep `report_array` as a class member because the state is in the database.

It looks like it can be a local variable in this function and calls to `report_array.clear()` would change to `storage->deleteReportEvents(...)`.

IMO, one of the important thing here is to try to minimize double-sent events. It looks like the backend should be able to discard duplicates by looking at the random uuid we send, though we should probably check with them.

Reply:
I modified the report_array. But I'm still thinking about how to prevent double-sent. Do we have a mechanism to check it with backend ? ( And does a check better than just send it ? )

Reply:
Checking before sending sounds complicated and not worth the overhead anyway. I'm trying to find out with the backend team if they can live with occasional resends.